bot/exts/info/doc/_parsing.py

```diff
@@ -0,0 +1,243 @@
+from __future__ import annotations
+
+import logging
+import re
+import string
+import textwrap
+from collections import namedtuple
+from typing import Collection, Iterable, Iterator, List, Optional, TYPE_CHECKING, Union
+
+from bs4 import BeautifulSoup
+from bs4.element import NavigableString, Tag
+
+from bot.utils.helpers import find_nth_occurrence
+from . import MAX_SIGNATURE_AMOUNT
+from ._html import get_dd_description, get_general_description, get_signatures
+from ._markdown import DocMarkdownConverter
+if TYPE_CHECKING:
+    from ._cog import DocItem
+
+log = logging.getLogger(__name__)
+
+_WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)(\s+)")
+_PARAMETERS_RE = re.compile(r"\((.+)\)")
+
+_NO_SIGNATURE_GROUPS = {
+    "attribute",
+    "envvar",
+    "setting",
+    "tempaltefilter",
+    "templatetag",
+    "term",
+}
+_EMBED_CODE_BLOCK_LINE_LENGTH = 61
+# _MAX_SIGNATURE_AMOUNT code block wrapped lines with py syntax highlight
+_MAX_SIGNATURES_LENGTH = (_EMBED_CODE_BLOCK_LINE_LENGTH + 8) * MAX_SIGNATURE_AMOUNT
+# Maximum discord message length - signatures on top - space for footer
+_MAX_DESCRIPTION_LENGTH = 1900 - _MAX_SIGNATURES_LENGTH
+_TRUNCATE_STRIP_CHARACTERS = "!?:;." + string.whitespace
+
+BracketPair = namedtuple("BracketPair", ["opening_bracket", "closing_bracket"])
+_BRACKET_PAIRS = {
+    "{": BracketPair("{", "}"),
+    "(": BracketPair("(", ")"),
+    "[": BracketPair("[", "]"),
+    "<": BracketPair("<", ">"),
+}
+
+
+def _is_closing_quote(search_string: str, index: int) -> bool:
+    """Check whether the quote at `index` inside `search_string` can be a closing quote."""
+    if search_string[index - 1] != "\\":
+        return True
+    elif search_string[index - 2] == "\\":
+        return True
+    return False
+
+
+def _split_parameters(parameters_string: str) -> Iterator[str]:
+    """
+    Split parameters of a signature into individual parameter strings on commas.
+
+    Long string literals are not accounted for.
+    """
```

Comment:
Optional arguments in brackets prevent this function from doing much in a lot of cases as it won't be able to split in them. Do you think it's worth looking into more? So far I haven't been able to come up with a satisfactory way to handle it later on in `_truncate_signatures` as it'll need to know the depth of optional brackets to be able to add them when not taking in all params

Reply:
We can be sure that they're at the end, there may be more levels of them though. Parsing them for enabling splits inside is fairly easy as other brackets will have a `=` somewhere between them and the last comma, but I couldn't think of a nice way to then add them at the end when truncating without some ugly counting going on or something like that.
Come to think of it I don't recall seeing any symbols with a long signature and a optionals, as those are done for functions from extension modules. I'll probably just leave it as it is unless I stumble on some real data that could benfit; the truncating and everything around it is a bit ugly already without this

Reply:
Yeah, it's a bit involved, but I don't envision it being _too_ ugly. It could be a done with a recursive parser that returns a flattened list of arguments while maintaining their opening/closing brackets, if present. A flat list is simple to truncate, but you do have to be mindful of the current depth of brackets at the truncation boundary so that you can insert the appropriate ellipses and closing brackets.

Reply:
Maybe making the parser return the text and depth as a tuple would make it cleaner. Or maybe not.